### PR TITLE
base-files: lint shell scripts

### PR DESCRIPTION
**Pull request by aparcar**
This patch lints all bash files installed via the base-files package.
Linting is a baseline for pretty code and shouldn't steal time away from
reviewers.

The tool `shfmt`[1] is used with options to fit the current style as good
as possible, however fixes up areas where it's not.

    shfmt -w -sr -s <filename>

Once existing, this command should be added to a CI.

To improve readability of JSON/jshn.sh functions every case of
`json_select ..` should be followed by a comment describing the new
level, e.g. `json_select .. # root`.

[1]: https://github.com/mvdan/sh

Suggested-by: Mariana Moreira <m.camargomoreira@gmail.com>
Signed-off-by: Paul Spooren <mail@aparcar.org>

### CODE REVIEW

**Comment by mpratt14** [comment]
it's interesting how this tool prefers piping `|` and command combination `&&` to go to the next line...

to me, it looks strange to have any line start with these symbols, and `&& \` looks much cleaner

are there any settings regarding this for that style tool?

**Comment by lynxis** [comment]
@aparcar great idea! I would love to see another PR adding this a CI job.

File: package/base-files/files/bin/config_generate [hunk 1]
```diff
@@ -95,10 +95,10 @@ generate_network() {
...
 
 	json_select network
-		json_select "$1"
-			json_get_vars device macaddr protocol ipaddr netmask vlan
-			json_get_values ports ports
-		json_select ..
+	json_select "$1"
+	json_get_vars device macaddr protocol ipaddr netmask vlan
+	json_get_values ports ports
+	json_select ..
```
Comment (musashino205) [inline package/base-files/files/bin/config_generate:101]: ```json_select``` moves current level to the specified object/array. I personally prefer the code with indentation before this change.

**Comment by dangowrt** [comment]
> it's interesting how this tool prefers piping `|` and command combination `&&` to go to the next line...
> 
> to me, it looks strange to have any line start with these symbols, and `&& \` looks much cleaner

while it may not be what most of us are used to, thinking about it, having those symbols at the beginning of the line makes things more easy to read, especially also for people relying on large fonts (as lines would get truncated unless you horizontal-scroll) or braille. so maybe we should actually change the style in existing scripts to that rather than telling the linter to ignore it.

**Comment by mpratt14** [comment]
> rather than telling the linter to ignore it.

I was wrong about that part, making it put | and && on new lines is an *extra* option

it's not critical of course, but this styling change makes the diff much larger

**Comment by aparcar** [comment]
I'm fine with either solution. Newlines before && improve readability from my point of view but it's really a taste. Since it's not default more people seem to prefer the && newline approach.

**Comment by neheb** [comment]
I still recommend manually moving them to a single line.

File: package/base-files/files/sbin/wifi [hunk 2]
```diff
@@ -194,8 +193,8 @@ set_wifi_down() {
 	local cfg="$1"
 	local vifs vif vifstr
 
-	[ -f "/var/run/wifi-${cfg}.pid" ] &&
-		kill "$(cat "/var/run/wifi-${cfg}.pid")" 2>/dev/null
+	[ -f "/var/run/wifi-${cfg}.pid" ] \
+		&& kill "$(cat "/var/run/wifi-${cfg}.pid")" 2> /dev/null
```
Comment (neheb) [inline package/base-files/files/sbin/wifi:197]: useless cat here. kill can read from stdin.

File: package/base-files/files/sbin/wifi [hunk 3]
```diff
@@ -194,8 +193,8 @@ set_wifi_down() {
 	local cfg="$1"
 	local vifs vif vifstr
 
-	[ -f "/var/run/wifi-${cfg}.pid" ] &&
-		kill "$(cat "/var/run/wifi-${cfg}.pid")" 2>/dev/null
+	[ -f "/var/run/wifi-${cfg}.pid" ] \
```
Comment (neheb) [inline package/base-files/files/sbin/wifi:196]: I'd merge both lines.

File: package/base-files/files/sbin/wifi [hunk 4]
```diff
@@ -151,37 +150,37 @@ wifi_reload() {
...
 }
 
-start_net() {(
+start_net() { (
 	local iface="$1"
 	local config="$2"
 	local vifmac="$3"
 
-	[ -f "/var/run/$iface.pid" ] && kill "$(cat /var/run/${iface}.pid)" 2>/dev/null
+	[ -f "/var/run/$iface.pid" ] && kill "$(cat /var/run/${iface}.pid)" 2> /dev/null
```
Comment (neheb) [inline package/base-files/files/sbin/wifi:175]: useless cat here. Just replace with

```
kill < "/var/run/${iface}.pid" 
```

File: package/base-files/files/sbin/wifi [hunk 5]
```diff
@@ -114,7 +113,7 @@ _wifi_updown() {
 			set disable
 		}
 		config_get iftype "$device" type
-		if eval "type ${1}_$iftype" 2>/dev/null >/dev/null; then
+		if eval "type ${1}_$iftype" 2> /dev/null > /dev/null; then
```
Comment (neheb) [inline package/base-files/files/sbin/wifi:116]: Should be replaced with

```
if eval "type ${1}_$iftype" 2>&1 /dev/null; then
```

File: package/base-files/files/sbin/wifi [hunk 6]
```diff
@@ -54,15 +53,15 @@ bridge_interface() {(
...
 
 prepare_key_wep() {
 	local key="$1"
 	local hex=1
 
 	echo -n "$key" | grep -qE "[^a-fA-F0-9]" && hex=0
-	[ "${#key}" -eq 10 -a $hex -eq 1 ] || \
-	[ "${#key}" -eq 26 -a $hex -eq 1 ] || {
+	[ "${#key}" -eq 10 -a $hex -eq 1 ] \
+		|| [ "${#key}" -eq 26 -a $hex -eq 1 ] || {
```
Comment (neheb) [inline package/base-files/files/sbin/wifi:64]: Move to one line

File: package/base-files/files/sbin/sysupgrade [hunk 7]
```diff
@@ -161,35 +175,37 @@ add_overlayfiles() {
...
-		list_static_conffiles | sort -u |
-			grep -h -v -x -F -f $conffiles > "$keepfiles"
+		list_static_conffiles | sort -u \
+			| grep -h -v -x -F -f $conffiles > "$keepfiles"
 
 		# backup conffiles, but only those changed if '-u'
-		[ $SKIP_UNCHANGED = 1 ] &&
-			list_changed_conffiles | sort -u > "$conffiles"
+		[ $SKIP_UNCHANGED = 1 ] \
+			&& list_changed_conffiles | sort -u > "$conffiles"
```
Comment (neheb) [inline package/base-files/files/sbin/sysupgrade:183]: Single line.

File: package/base-files/files/sbin/sysupgrade [hunk 8]
```diff
@@ -161,35 +175,37 @@ add_overlayfiles() {
 
 		# backup files from /etc/sysupgrade.conf and /lib/upgrade/keep.d, but
 		# ignore those aready controlled by opkg conffiles
-		list_static_conffiles | sort -u |
-			grep -h -v -x -F -f $conffiles > "$keepfiles"
+		list_static_conffiles | sort -u \
+			| grep -h -v -x -F -f $conffiles > "$keepfiles"
```
Comment (neheb) [inline package/base-files/files/sbin/sysupgrade:179]: Single line.

File: package/base-files/files/sbin/sysupgrade [hunk 9]
```diff
@@ -135,15 +146,18 @@ list_static_conffiles() {
...
 add_conffiles() {
 	local file="$1"
 
-	( list_static_conffiles "$find_filter"; list_changed_conffiles ) |
-		sort -u > "$file"
+	(
+		list_static_conffiles "$find_filter"
+		list_changed_conffiles
+	) \
+		| sort -u > "$file"
```
Comment (neheb) [inline package/base-files/files/sbin/sysupgrade:160]: Move up to the same line as )

File: package/base-files/files/sbin/led.sh [hunk 10]
```diff
@@ -10,16 +10,15 @@ do_led() {
 	config_get name $1 name
 	config_get sysfs $1 sysfs
 	[ "$name" = "$NAME" -o "$sysfs" = "$NAME" -a -e "/sys/class/leds/${sysfs}" ] && {
-		[ "$ACTION" = "set" ] &&
-			echo 1 >/sys/class/leds/${sysfs}/brightness \
-			|| echo 0 >/sys/class/leds/${sysfs}/brightness
+		[ "$ACTION" = "set" ] \
+			&& echo 1 > /sys/class/leds/${sysfs}/brightness \
```
Comment (neheb) [inline package/base-files/files/sbin/led.sh:14]: Hmm this is an if statement in disguise. It is possible to simplify?

File: package/base-files/files/lib/upgrade/stage2 [hunk 11]
```diff
@@ -149,7 +148,7 @@ sleep 6
 
 echo 3 > /proc/sys/vm/drop_caches
 
-if [ -n "$IMAGE" ] && type 'platform_pre_upgrade' >/dev/null 2>/dev/null; then
+if [ -n "$IMAGE" ] && type 'platform_pre_upgrade' > /dev/null 2> /dev/null; then
```
Comment (neheb) [inline package/base-files/files/lib/upgrade/stage2:151]: 2>&1 /dev/null

File: package/base-files/files/lib/upgrade/stage2 [hunk 12]
```diff
@@ -126,20 +125,20 @@ while read -r a b c; do
 		MemT*) mem="$b" ;; esac
 done < /proc/meminfo
 
-[ "$mem" -gt 32768 ] && \
-	skip_services="dnsmasq log network"
+[ "$mem" -gt 32768 ] \
+	&& skip_services="dnsmasq log network"
```
Comment (neheb) [inline package/base-files/files/lib/upgrade/stage2:129]: single line

File: package/base-files/files/lib/upgrade/nand.sh [hunk 13]
```diff
@@ -98,10 +97,10 @@ identify_tar() {
 
 nand_restore_config() {
 	sync
-	local ubidev=$( nand_find_ubi $CI_UBIPART )
-	local ubivol="$( nand_find_volume $ubidev rootfs_data )"
-	[ ! "$ubivol" ] &&
-		ubivol="$( nand_find_volume $ubidev $CI_ROOTPART )"
+	local ubidev=$(nand_find_ubi $CI_UBIPART)
+	local ubivol="$(nand_find_volume $ubidev rootfs_data)"
+	[ ! "$ubivol" ] \
```
Comment (neheb) [inline package/base-files/files/lib/upgrade/nand.sh:102]: sounds like ! can be removed and && replaced with ||.

File: package/base-files/files/lib/upgrade/do_stage2 [hunk 14]
```diff
@@ -5,13 +5,13 @@
 include /lib/upgrade
 
 v "Performing system upgrade..."
-if type 'platform_do_upgrade' >/dev/null 2>/dev/null; then
+if type 'platform_do_upgrade' > /dev/null 2> /dev/null; then
```
Comment (neheb) [inline package/base-files/files/lib/upgrade/do_stage2:8]: 2>&1 /dev/null

File: package/base-files/files/lib/preinit/30_failsafe_wait [hunk 15]
```diff
@@ -32,8 +32,8 @@ fs_wait_for_key () {
 		while [ $timer -gt 0 ]; do
 			pi_failsafe_net_message=true \
 				preinit_net_echo "Please press button now to enter failsafe"
-			echo "$timer" >$keypress_sec
-			timer=$(($timer - 1))
+			echo "$timer" > $keypress_sec
+			timer=$((timer - 1))
```
Comment (neheb) [inline package/base-files/files/lib/preinit/30_failsafe_wait:36]: nice. I assume the automatic $ removal in (()) is automatic?

File: package/base-files/files/etc/protocols [hunk 16]
```diff
@@ -6,52 +6,51 @@
 # assigned by IANA and are not historical.
 # If you need a huge list of used numbers please install the nmap package.
 
-ip	0	IP		# internet protocol, pseudo protocol number
+ip 0 IP # internet protocol, pseudo protocol number
 #hopopt	0	HOPOPT		# IPv6 Hop-by-Hop Option [RFC1883]
-icmp	1	ICMP		# internet control message protocol
```
Comment (neheb) [inline package/base-files/files/etc/protocols:11]: This is an sh file?

**Comment by neheb** [comment]
Fun idea: let's take this to the packages feed. There's a lot of opportunity there. That's where the playground is anyway...

**Comment by aparcar** [comment]
Sure I'll add it to the CI

File: package/base-files/files/sbin/sysupgrade (hunk 8 above)
Comment (aparcar) [reply package/base-files/files/sbin/sysupgrade:179]: I don't want to exceed 88 chars per line.

File: package/base-files/files/sbin/sysupgrade (hunk 7 above)
Comment (aparcar) [reply package/base-files/files/sbin/sysupgrade:183]: I don't want to exceed 88 chars per line.

File: package/base-files/files/lib/preinit/30_failsafe_wait (hunk 15 above)
Comment (aparcar) [reply package/base-files/files/lib/preinit/30_failsafe_wait:36]: Correct

File: package/base-files/files/etc/protocols (hunk 16 above)
Comment (aparcar) [reply package/base-files/files/etc/protocols]: Accident, removed.

File: package/base-files/files/lib/upgrade/do_stage2 (hunk 14 above)
Comment (aparcar) [reply package/base-files/files/lib/upgrade/do_stage2:8]: This should be replaced my `2>&1 > /dev/null` right?

File: package/base-files/files/sbin/wifi (hunk 2 above)
Comment (aparcar) [reply package/base-files/files/sbin/wifi:196]: Shellcheck tells me it can not.

Comment (neheb) [reply package/base-files/files/sbin/wifi:196]: It doesn’t warn on it but kill definitely accepts stdin.

Comment (neheb) [reply package/base-files/files/sbin/wifi:196]: I think shellcheck warns only on shell builtins.

File: package/base-files/files/lib/upgrade/do_stage2 (hunk 14 above)
Comment (neheb) [reply package/base-files/files/lib/upgrade/do_stage2]: Yeah I think so.

File: package/base-files/files/sbin/sysupgrade (hunk 7 above)
Comment (aparcar) [reply package/base-files/files/sbin/sysupgrade:183]: Just as a clarification I'm not particularly interested in the number, it's just what the Python linter `black` uses as default.

File: package/base-files/files/sbin/wifi (hunk 2 above)
Comment (aparcar) [reply package/base-files/files/sbin/wifi:196]:
```
kill < pid
ash: you need to specify whom to kill
```
Are you sure?

File: package/base-files/files/etc/services [hunk 17]
```diff
@@ -1,173 +1,172 @@
...
-ospfd		2604/tcp
-bgpd		2605/tcp
-ospf6d		2606/tcp
-ospfapi		2607/tcp
-isisd		2608/tcp
-sane-port	6566/tcp	sane saned
-ircd		6667/tcp
-git		9418/tcp
-
+echo 7/tcp
```
Comment (neheb) [inline package/base-files/files/etc/services:1]: This is not a shell file.

File: package/base-files/files/lib/functions/network.sh [hunk 18]
```diff
@@ -17,9 +17,9 @@ __network_ifstatus() {
 
 	__tmp="$(jsonfilter ${4:+-F "$4"} ${5:+-l "$5"} -s "${__NETWORK_CACHE:-{}}" -e "$1=@.interface${2:+[@.interface='$2']}$3")"
 
-	[ -z "$__tmp" ] && \
-		unset "$1" && \
-		return 1
+	[ -z "$__tmp" ] \
```
Comment (neheb) [inline package/base-files/files/lib/functions/network.sh:20]: One line I think.

Comment (aparcar) [reply package/base-files/files/lib/functions/network.sh:20]: The condition has the first line, all subsequent things should have their own lines.

File: package/base-files/files/lib/preinit/02_sysinfo [hunk 19]
```diff
@@ -1,10 +1,10 @@
...
 	[ -d /proc/device-tree ] || return
 	mkdir -p /tmp/sysinfo
-	[ -e /tmp/sysinfo/board_name ] || \
-		echo "$(strings /proc/device-tree/compatible | head -1)" > /tmp/sysinfo/board_name
-	[ ! -e /tmp/sysinfo/model -a -e /proc/device-tree/model ] && \
-		echo "$(cat /proc/device-tree/model)" > /tmp/sysinfo/model
+	[ -e /tmp/sysinfo/board_name ] \
+		|| echo "$(strings /proc/device-tree/compatible | head -1)" > /tmp/sysinfo/board_name
+	[ ! -e /tmp/sysinfo/model -a -e /proc/device-tree/model ] \
+		&& echo "$(cat /proc/device-tree/model)" > /tmp/sysinfo/model
```
Comment (neheb) [inline package/base-files/files/lib/preinit/02_sysinfo:7]: sounds like a useless echo

Comment (aparcar) [reply package/base-files/files/lib/preinit/02_sysinfo]: True, just like `strings` before that

File: package/base-files/files/sbin/wifi (hunk 2 above)
Comment (neheb) [reply package/base-files/files/sbin/wifi:196]: Huh. Just tested it. It seems kill doesn't actually work with stdin. Wonder how I got it working...

File: package/base-files/files/lib/upgrade/nand.sh (hunk 13 above)
Comment (aparcar) [reply package/base-files/files/lib/upgrade/nand.sh:102]: wouldn't that require e.g. -n?

Comment (neheb) [reply package/base-files/files/lib/upgrade/nand.sh:102]: No idea what you mean. It's just an if inversion.

```
[ "$ubivol" ] || ubivol="$(nand_find_volume $ubidev $CI_ROOTPART)"
```

Comment (aparcar) [reply package/base-files/files/lib/upgrade/nand.sh:102]: Don't we need `[ -n "$ubivol ]`?

Comment (neheb) [reply package/base-files/files/lib/upgrade/nand.sh:102]: Hmm I think you're right. Alternatively, -z &&.

Comment (aparcar) [reply package/base-files/files/lib/upgrade/nand.sh:102]: :woman_dancing:

Comment (mpratt14) [reply package/base-files/files/lib/upgrade/nand.sh:102]: the string by itself `[ "$ubivol" ]` is the same as -n `[ -n "$ubivol" ]`

Comment (adschm) [reply package/base-files/files/lib/upgrade/nand.sh:102]: well, is it really the same? Have you checked all cases of "", undefined, whatever?

And if it really was, wouldn't -n still be much easier to understand and it's behavior be more well-known by most people?

File: package/base-files/files/etc/hotplug.d/net/00-sysctl [hunk 20]
```diff
@@ -2,8 +2,8 @@
 
 if [ "$ACTION" = add ]; then
 	for CONF in /etc/sysctl.d/*.conf /etc/sysctl.conf; do
-		[ ! -f "$CONF" ] && continue;
-		sed -ne "/^[[:space:]]*net\..*\.$DEVICENAME\./p" "$CONF" | \
-			sysctl -e -p - | logger -t sysctl
+		[ ! -f "$CONF" ] && continue
+		sed -ne "/^[[:space:]]*net\..*\.$DEVICENAME\./p" "$CONF" \
+			| sysctl -e -p - | logger -t sysctl
```
Comment (adschm) [inline package/base-files/files/etc/hotplug.d/net/00-sysctl:7]: That's against our common practice.
I also personally prefer the old version.

File: package/base-files/files/etc/init.d/led [hunk 21]
```diff
@@ -53,70 +53,70 @@ load_led() {
 			"$sysfs" \
 			"$(sed -ne 's/^.*\[\(.*\)\].*$/\1/p' /sys/class/leds/${sysfs}/trigger)" \
 			"$(cat /sys/class/leds/${sysfs}/brightness)" \
-				>> /var/run/led.state
+			>> /var/run/led.state
 
-		[ "$default" = 0 ] &&
-			echo 0 >/sys/class/leds/${sysfs}/brightness
+		[ "$default" = 0 ] \
+			&& echo 0 > /sys/class/leds/${sysfs}/brightness
```
Comment (adschm) [inline package/base-files/files/etc/init.d/led:59]: See above. I'm not a fan of this wrapping logic, and, more importantly, it's against the style we use everywhere else.

There are similar cases below which I won't highlight anymore ...

File: package/base-files/files/etc/iproute2/rt_tables [hunk 22]
```diff
@@ -1,11 +1,11 @@
...
-128	prelocal
-255	local
-254	main
-253	default
-0	unspec
+128 prelocal
+255 local
+254 main
+253 default
+0 unspec
```
Comment (adschm) [inline package/base-files/files/etc/iproute2/rt_tables:8]: I'd remove this change.

File: package/base-files/files/etc/profile [hunk 23]
```diff
@@ -15,7 +15,7 @@ export PS1='\u@\h:\w\$ '
 export ENV=/etc/shinit
 
 case "$TERM" in
-	xterm*|rxvt*)
+	xterm* | rxvt*)
```
Comment (adschm) [inline package/base-files/files/etc/profile:18]: I'd be interested in the reason for that ...

File: package/base-files/files/etc/profile [hunk 24]
```diff
@@ -27,10 +27,9 @@ esac
 	unset FILE
 }
 
-if ( grep -qs '^root::' /etc/shadow && \
-     [ -z "$FAILSAFE" ] )
-then
-cat << EOF
+if (grep -qs '^root::' /etc/shadow \
+	&& [ -z "$FAILSAFE" ]); then
```
Comment (adschm) [inline package/base-files/files/etc/profile:31]: This could actually fit into one line ...

**Comment by adschm** [comment]
I stopped reviewing at about one third. Most of the stuff is certainly valid and good to have. Some stuff is taste (e.g. whether we really need to change the indent on all these select-cases, as our formatting is perfectly readable and consistent at least locally as well). And there are a few things which actually make it worse in my opinion (e.g. the line wrapping for pipe and logic (&& ||) operators.
I'm personally agnostic about the json_select indent. However, note that strictly that would mean we keep the old version as we have no strong reason to _change_ the existing style.
And I don't have any idea about the space before EOF ...

I haven't read the discussion above, but I assume there are options where part of the changes can be disabled. I'd reduce this to the changes we can all agree on. In any case, in the first third I've looked at there has not been any "important" issue, so strictly this whole patch is not really needed at all, or maybe the package wasn't a good showcase ;-)

**Comment by mpratt14** [comment]
Another fun idea, lets make our own linter as a script (python if bash becomes too difficult)

**Comment by neheb** [comment]
> 
> 
> Another fun idea, lets make our own linter as a script (python if bash becomes too difficult)

You're welcome to do so if you feel shfmt is not good enough.

**Comment by aparcar** [comment]
@adschm 

> Some stuff is taste (e.g. whether we really need to change the indent on all these select-cases, as our formatting is perfectly readable and consistent at least locally as well)

I'll get rid of the case indent thing, I thought I'd be preferable but it's neither default of `shfmt` nor super the default in our code.

> And there are a few things which actually make it worse in my opinion (e.g. the line wrapping for pipe and logic (&& ||) operators.

If you don't mind I still opt for this as it seemingly improves readability for people using Braille or text2speech readers.

> I'm personally agnostic about the json_select indent. However, note that strictly that would mean we keep the old version as we have no strong reason to change the existing style.

The *strong* reason is automation and `shfmt` doesn't understand what our JSON lib does.

On the other hand we can trivially work around this by using comments, similar as the comments used in Makefile.

```
        json_select network                                                             
        json_select "lan"                                                               
        json_get_vars device                                                            
        json_get_values ports ports                                                     
        json_select .. # lan                                                            
        json_select .. # network   
```

> And I don't have any idea about the space before EOF ...

It's the default of a popular tool used by plenty of people, I wouldn't spend time on arguing about this to much.

> In any case, in the first third I've looked at there has not been any "important" issue, so strictly this whole patch is not really needed at all, or maybe the package wasn't a good showcase ;-)

This patch looks at style not mechanics. Another patch iteration could apply `shellcheck` to everything. Since this patch "fixes" plenty of style inconsistencies I think it's a ok showcase.

**Comment by aparcar** [comment]
> Another fun idea, lets make our own linter as a script (python if bash becomes too difficult)

Please don't pursue this idea (within this thread). I have no intentions to become the author/maintainer of yet another thing if a good (open source) solution already exists :)

**Comment by aparcar** [comment]
@adschm I tried again without the case indent and even more lines changed :)

**Comment by mpratt14** [comment]
> Please don't pursue this idea (within this thread)

of course, it would be my own time, if I could find enough free time to play with that :)

unless someone else is also interested

**Comment by aparcar** [comment]
This is long outdated.

**Pull request closed without merging**